File: dungeons-1.21.11/src/main/java/fishmod/utils/dungeon/Split.java
package fishmod.utils.dungeon;

import fishmod.utils.Constants;
import config.practical.manager.ConfigValue;
import net.minecraft.client.font.TextRenderer;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.text.MutableText;
import net.minecraft.text.Text;

public class Split {

    public enum TimerType {
        TICK_TIME("Tick time"), DIFFRENCE("difference");

        private final String label;

        TimerType(String label) {
            this.label = label;
        }

        @Override
        public String toString() {
            return label;
        }
    }

    public static final int GREEN = 5635925;
    public static final int GRAY = 11184810;
    public static final int DARK_GRAY = 5592405;

    @ConfigValue
    public static int realTimeColorInactive = GREEN;
    @ConfigValue
    public static int realTimeColorOngoing = GREEN;
    @ConfigValue
    public static int realTimeColorComplete = GREEN;

    @ConfigValue
    public static int serverTimeColorInactive = GRAY;
    @ConfigValue
    public static int serverTimeColorOngoing = GRAY;
    @ConfigValue
    public static int serverTimeColorComplete = GRAY;

    @ConfigValue
    public static int parenthesesColorInactive = DARK_GRAY;
    @ConfigValue
    public static int parenthesesColorOngoing = DARK_GRAY;
    @ConfigValue
    public static int parenthesesColorComplete = DARK_GRAY;

    @ConfigValue
    public static TimerType timerType = TimerType.TICK_TIME;

    private final String name, startString, endString;
    private final int color;
    private final double avg;
    private int tick;
    private long startTime, endTime;
    private boolean started, ended;

    public Split(String name, String startString, String endString, int color, double avg) {
        this.name = name;
        this.startString = startString;
        this.endString = endString;
        this.color = color;
        this.avg = avg;
        this.tick = 0;
        this.ended = false;
    }

    public double getAvg() {
        return avg;
    }

    public void parseMessage(String string) {
        if (!started) {
            if (startString.equals(string)) {
                start();
            }
        } else if (!ended) {
            if (endString.equals(string)) {
                end();
            }
        }
    }

    public void tick() {
        if (started && !ended) {
            this.tick++;
        }
    }

    public void reset() {
        tick = 0;
        ended = false;
        started = false;
    }

    public void end() {
        if (ended) return;
        endTime = System.currentTimeMillis();
        started = false;
        ended = true;
    }

    public void start() {
        startTime = System.currentTimeMillis();
        started = true;
    }

    public String getName() {
        return name;
    }

    public boolean started() {
        return started;
    }

    public boolean ended() {
        return ended;
    }

    public double getTickTime() {
        return tick * Constants.TICK_DURATION;
    }

    public double getRealTime() {
        // startTime == 0 means start() was never called; guard against returning
        // epoch-time-in-seconds when printSplits() force-ends a never-started split.
        if (startTime == 0) return 0;
        double realTime;
        if (ended) {
            realTime = (endTime - startTime) / 1000.0;
        } else if (started) {
            realTime = (System.currentTimeMillis() - startTime) / 1000.0;
        } else {
            realTime = 0;
        }
        return realTime;
    }

    public MutableText createNameText() {
        return Text.literal(name + " ").withColor(color);
    }

    public double getTimeDiffrence() {
        return getRealTime() - getTickTime();
    }

    public MutableText createTimeText() {
        int realTimeColor, serverTimeColor, parenthesesColor;

        if (!started) {
            realTimeColor = realTimeColorInactive;
            serverTimeColor = serverTimeColorInactive;
            parenthesesColor = parenthesesColorInactive;
        } else if (!ended) {
            realTimeColor = realTimeColorOngoing;
            serverTimeColor = serverTimeColorOngoing;
            parenthesesColor = parenthesesColorOngoing;
        } else {
            realTimeColor = realTimeColorComplete;
            serverTimeColor = serverTimeColorComplete;
            parenthesesColor = parenthesesColorComplete;
        }

        double tickTime = getTickTime();
        double realTime = getRealTime();

        String serverTime;
        if (timerType == TimerType.DIFFRENCE) {
            double diff = realTime - tickTime;
            if (diff > 0) {
                serverTime = "+" + Constants.DECIMAL_FORMAT.format(diff) + "s";
            } else {
                serverTime = Constants.DECIMAL_FORMAT.format(diff) + "s";
            }
        } else {
            //default to tick timer
            serverTime = Constants.DECIMAL_FORMAT.format(tickTime) + "s";
        }

        String realTimeString = (realTime >= 60? (int)(realTime / 60) + "m ": "") + Constants.DECIMAL_FORMAT.format(realTime % 60) + "s";
        return Text.literal(realTimeString).withColor(realTimeColor)
                .append(Text.literal(" (").withColor(parenthesesColor)
                        .append(Text.literal(serverTime).withColor(serverTimeColor))
                        .append(Text.literal(")").withColor(parenthesesColor)
                        ));
    }

    public void drawSplit(DrawContext context, TextRenderer textRenderer, int x, int y, int maxWidth) {
        Text nameText = createNameText();
        Text timerText = createTimeText();

        int timerWidth = textRenderer.getWidth(timerText);
        context.drawText(textRenderer, nameText, x, y, 0xffffffff, true);
        context.drawText(textRenderer, timerText, x + maxWidth - timerWidth, y, 0xffffffff, true);
    }
}

File: dungeons-1.21.11/src/main/kotlin/fishmod/utils/dungeon/Split.kt
package fishmod.utils.dungeon

import config.practical.manager.ConfigValue
import fishmod.utils.Constants
import net.minecraft.client.font.TextRenderer
import net.minecraft.client.gui.DrawContext
import net.minecraft.text.MutableText
import net.minecraft.text.Text

class Split(
    val name: String,
    private val startString: String,
    private val endString: String,
    private val color: Int,
    val avg: Double
) {

    enum class TimerType(private val label: String) {
        TICK_TIME("Tick time"), DIFFRENCE("difference");

        override fun toString(): String = label
    }

    companion object {
        const val GREEN: Int = 5635925
        const val GRAY: Int = 11184810
        const val DARK_GRAY: Int = 5592405

        @ConfigValue
        @JvmField
        var realTimeColorInactive: Int = GREEN
        @ConfigValue
        @JvmField
        var realTimeColorOngoing: Int = GREEN
        @ConfigValue
        @JvmField
        var realTimeColorComplete: Int = GREEN

        @ConfigValue
        @JvmField
        var serverTimeColorInactive: Int = GRAY
        @ConfigValue
        @JvmField
        var serverTimeColorOngoing: Int = GRAY
        @ConfigValue
        @JvmField
        var serverTimeColorComplete: Int = GRAY

        @ConfigValue
        @JvmField
        var parenthesesColorInactive: Int = DARK_GRAY
        @ConfigValue
        @JvmField
        var parenthesesColorOngoing: Int = DARK_GRAY
        @ConfigValue
        @JvmField
        var parenthesesColorComplete: Int = DARK_GRAY

        @ConfigValue
        @JvmField
        var timerType: TimerType = TimerType.TICK_TIME
    }

    private var tick: Int = 0
    private var startTime: Long = 0
    private var endTime: Long = 0
    private var started: Boolean = false
    private var ended: Boolean = false

    fun parseMessage(string: String) {
        if (!started) {
            if (startString == string) {
                start()
            }
        } else if (!ended) {
            if (endString == string) {
                end()
            }
        }
    }

    fun tick() {
        if (started && !ended) {
            tick++
        }
    }

    fun reset() {
        tick = 0
        ended = false
        started = false
    }

    fun end() {
        if (ended) return
        endTime = System.currentTimeMillis()
        started = false
        ended = true
    }

    fun start() {
        startTime = System.currentTimeMillis()
        started = true
    }

    fun started(): Boolean = started

    fun ended(): Boolean = ended

    fun getTickTime(): Double = tick * Constants.TICK_DURATION

    fun getRealTime(): Double {
        // startTime == 0 means start() was never called; guard against returning
        // epoch-time-in-seconds when printSplits() force-ends a never-started split.
        if (startTime == 0L) return 0.0
        return if (ended) {
            (endTime - startTime) / 1000.0
        } else if (started) {
            (System.currentTimeMillis() - startTime) / 1000.0
        } else {
            0.0
        }
    }

    fun createNameText(): MutableText = Text.literal("$name ").withColor(color)

    fun getTimeDiffrence(): Double = getRealTime() - getTickTime()

    fun createTimeText(): MutableText {
        val realTimeColor: Int
        val serverTimeColor: Int
        val parenthesesColor: Int

        if (!started) {
            realTimeColor = realTimeColorInactive
            serverTimeColor = serverTimeColorInactive
            parenthesesColor = parenthesesColorInactive
        } else if (!ended) {
            realTimeColor = realTimeColorOngoing
            serverTimeColor = serverTimeColorOngoing
            parenthesesColor = parenthesesColorOngoing
        } else {
            realTimeColor = realTimeColorComplete
            serverTimeColor = serverTimeColorComplete
            parenthesesColor = parenthesesColorComplete
        }

        val tickTime = getTickTime()
        val realTime = getRealTime()

        val serverTime: String = if (timerType == TimerType.DIFFRENCE) {
            val diff = realTime - tickTime
            if (diff > 0) {
                "+" + Constants.DECIMAL_FORMAT.format(diff) + "s"
            } else {
                Constants.DECIMAL_FORMAT.format(diff) + "s"
            }
        } else {
            // default to tick timer
            Constants.DECIMAL_FORMAT.format(tickTime) + "s"
        }

        val realTimeString = (if (realTime >= 60) (realTime / 60).toInt().toString() + "m " else "") + Constants.DECIMAL_FORMAT.format(realTime % 60) + "s"
        return Text.literal(realTimeString).withColor(realTimeColor)
            .append(
                Text.literal(" (").withColor(parenthesesColor)
                    .append(Text.literal(serverTime).withColor(serverTimeColor))
                    .append(Text.literal(")").withColor(parenthesesColor))
            )
    }

    fun drawSplit(context: DrawContext, textRenderer: TextRenderer, x: Int, y: Int, maxWidth: Int) {
        val nameText = createNameText()
        val timerText = createTimeText()

        val timerWidth = textRenderer.getWidth(timerText)
        context.drawText(textRenderer, nameText, x, y, -0x1, true)
        context.drawText(textRenderer, timerText, x + maxWidth - timerWidth, y, -0x1, true)
    }
}
